folding/src/examples/generic.rs

```diff
@@ -0,0 +1,180 @@
+use crate::{
+    expressions::{FoldingColumnTrait, FoldingCompatibleExpr, FoldingCompatibleExprInner},
+    FoldingConfig, Sponge,
+};
+use ark_ff::{Field, One, Zero};
+use kimchi::curve::KimchiCurve;
+use mina_poseidon::{
+    constants::PlonkSpongeConstantsKimchi,
+    sponge::{DefaultFqSponge, ScalarChallenge},
+    FqSponge,
+};
+use poly_commitment::PolyComm;
+use std::{
+    iter::successors,
+    rc::Rc,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+#[cfg(not(test))]
+use log::debug;
+#[cfg(test)]
+use std::println as debug;
+
+// 0. We start by defining the field and the curve that will be used in the
+// constraint system, in addition to the sponge that will be used to generate
+// challenges.
+pub type Fp = ark_bn254::Fr;
+pub type Curve = ark_bn254::G1Affine;
+pub type SpongeParams = PlonkSpongeConstantsKimchi;
+pub type BaseSponge = DefaultFqSponge<ark_bn254::g1::Parameters, SpongeParams>;
+
+// 1. We continue by defining a generic type of columns and selectors.
+// The selectors can be seen as additional (public) columns that are not part of
+// the witness.
+// The column must implement the trait [Hash] as it will be used by internal
+// structures of the library.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub enum Column {
+    X(usize),
+    Selector(usize),
+}
+
+// 2. We implement the trait [FoldingColumnTrait] that allows to distinguish
+// between the public and private inputs, often called the "instances" and the
+// "witnesses".
+// By default, we consider that the columns are all witness values and selectors
+// are public.
+impl FoldingColumnTrait for Column {
+    fn is_witness(&self) -> bool {
+        match self {
+            Column::X(_) => true,
+            Column::Selector(_) => false,
```

Comment:
Actually according to Fabrizio, the selectors we will be using in the real examples of MIPS and Keccak should have Selectors also set to true in terms of witness, but I guess for these toy examples it is fine to assume they are false.